### PR TITLE
Add support for new architecture (fabric)

### LINKED ISSUE #931: Lottie is autoplaying on iOS even if AutoPlay is turned to False
### Description

My lottie file autoplays directly when component mounts, even if auoplay is off. Only on iOS tho.
image

**Expected behavior:** to load the lottie without directly playing it

**Actual behavior:** it autoplays anywway

### Versions

"lottie-ios": "3.2.3",
"lottie-react-native": "^5.1.3",
"react-native": "^0.69.3",

### PR DESCRIPTION
**Pull request by AlexanderEggers**
This is the basic fabric integration for Android. I used a very basic fabric project to verify that the most basic props are working on both platforms - but that requires more testing to verify each and every permutation. 

### Changes made in this PR
- Fabric support for Android.
   - Note: Fabric support will only work for RN 0.70 and higher. That is due to the introduced autolinking which doesn't exists for older versions.
-  I migrated the existing project setup to use a monorepo setup instead to simplify some of the tooling usage.
- Added a new sample project to cover fabric testing
  - Note: The fabric-example project does not contain the windows source files including the react-native-windows package. I also had to use a much simpler example project implementation due to multiple libs not supporting fabric yet.
- Update project to support the latest RN 0.70.x
 - Refactor android project to allow old arch and new arch support. Essentially I created an implementation class that is called from both arch view manager files.
 - Update JS project to include the definitions needed for the RN codegen and relevant files to compile correctly.
 - Replace current output bundle creation with react-native-builder-bob.
 - Remove the android LottieView prop "cacheStrategy". That prop does not have a way to be injected into the lottie integration outside the XML definition.
 - Fix autoplay usage in the lottie view integration which was auto playing even if the prop was set to false. Potentially fixes #931.

### New project setup
You need to follow these steps to run the new monorepo structure:
1. Run `yarn install` in the root folder of the project to install all dependencies
2. Run `yarn setup` in the root folder of the project to build the library output files which are going to be consumed by third party projects (and examples).
3. Run the following scripts to run different parts of the project:
- `yarn fabric:android` - android example project running fabric
- `yarn fabric:ios` - ios example project running fabric
- `yarn paper:android` - android example project running paper
- `yarn paper:ios` - ios example project running paper

### CODE REVIEW

**Comment by AlexanderEggers** [comment]
@emilioicai @matinzd Can you have a look at the changes and the "Things to discuss" section? Just so that we can unblock the remaining work for this PR.

**Comment by matinzd** [comment]
> ### Things to discuss:
> * Please have a look at src/index.d.ts. I added a few comments pointing out certain props which cannot be supported by the new arch. I am actually confused how some of these were working in the first place because some were not even having a `@ReactProp` refering to them.
> * The color and text filter props are something which we need to implement differently. The new arch doesn't support the current complex objects we are attempting to send to the native side. Instead I would suggest we parse jsons from the JS to native side and use for example GSON (on Android) to get the objects again.
> 

- I also noticed that some props did not have `@ReactProp` annotation. 
- For the support of complex objects, serializing it can be a good idea but in that case, we will lose type safety. We can investigate what other libraries are doing. Maybe we can just have some kind of typescript interfaces that can be ignored but serialized under the hood.

File: package.json
Comment (capezzbr) [inline package.json:70]: I believe it would be better to support RN 0.68+ instead of jumping all the way to RN 0.70 (since it's not out yet)

Comment (matinzd) [reply package.json:70]: There are so many improvements to cli and codegen and bug fixes. Although this is just fabric example app, we need to keep it updated as much as we can to make sure everything is well aligned with the new architecture.

Comment (capezzbr) [reply package.json:70]: It would be awesome if you could at least try to support RN 0.69+, since otherwise adoption for this will be slow (e.g. folks will need to be on the latest version of RN, which isn't always possible right away).

From my understanding 0.70 brings auto-linking on Android, but that can be achieved via [this workaround](https://github.com/th3rdwave/react-native-safe-area-context/blob/d14de35ac6faa1dd61d370e04443999a86f2d5b2/android/src/main/jni/SafeAreaContextComponentsRegistry.cpp#L53-L59), or you can tell folks to manually link the module in their app `ComponentsRegistry.cpp` when using RN 0.69.

Unless 0.70 brings in some bugfix that does block this migration, and I'd be curious to learn more 🙌

Comment (matinzd) [reply package.json:70]: Fair point.

Comment (AlexanderEggers) [reply package.json:70]: You will be able to use the lib with any RN version that supports the new arch. We want require a certain, just recommend one.

**Comment by AlexanderEggers** [comment]
@emilioicai @matinzd FYI, I moved the project to yarn workspaces to simplify our dependency handling between the lib and the examples. Let me know what you think about that change.

Unfortunately I am still unable to solve the new architecture issue - I hope to get some help from the RN team in the next few weeks.

**Comment by matinzd** [comment]
> @emilioicai @matinzd FYI, I moved the project to yarn workspaces to simplify our dependency handling between the lib and the examples. Let me know what you think about that change.
> 
> Unfortunately I am still unable to solve the new architecture issue - I hope to get some help from the RN team in the next few weeks.

Thanks @AlexanderEggers ! I will check it out ASAP.

**Comment by AlexanderEggers** [comment]
@emilioicai @matinzd - can you have a look at the current changes in this branch? I am still blocked on that mentioned startup crash but I am hopening that maybe you can verify if my current setup looks fine and is not missing anything.

**Comment by emilioicai** [comment]
Hey @AlexanderEggers , sure, I'll have a look into it

**Comment by AlexanderEggers** [comment]
@emilioicai @matinzd I finally managed to get android working via fabric. 🎉  The animation seems to run fine and the basic props are working. I will need to figure out something for the color filter and text filter props. I will try to work on that in the next days. I will also need to work on verifying the paper architecture.

fabric-example

**Comment by AlexanderEggers** [comment]
@emilioicai @matinzd I already found some little time to also have a look at android paper and I was able to verify that paper ios and android is still working fine with the latest changes I made for android fabric. It would be great if you could review this PR. I think we are ready for an initial alpha version which just includes Android support for now. iOS should follow in the near future when I find enough time. Let me know if you have any questions/issues.

**Comment by AlexanderEggers** [comment]
@emilioicai I am unable to test if windows is still working fine with the latest js changes. Would you be able to do some testing around that?

**Comment by matinzd** [comment]
> @emilioicai @matinzd I already found some little time to also have a look at android paper and I was able to verify that paper ios and android is still working fine with the latest changes I made for android fabric. It would be great if you could review this PR. I think we are ready for an initial alpha version which just includes Android support for now. iOS should follow in the near future when I find enough time. Let me know if you have any questions/issues.

Thanks, @AlexanderEggers! 
I will try to check it out this week.

File: package.json
Comment (AlexanderEggers) [reply package.json]: @capezzbr FYI, for the initial version of fabric support for this lib, we will probably stick to only supporting 0.70.x and newer. If I happen to find some additional time in the future, I definitely will take care of adding support for 0.68.x and higher.

**Comment by AlexanderEggers** [comment]
@emilioicai @matinzd I just had a quick test with ios via the fabric example project and I actually have no issues compiling and running that project. I can see the animation in-app without changing a single line of code in the lottie package. It seems like that our current ios intergation is already supporting fabric without actually needing to change anything 🤔 It would great if someone could confirm that observation for me.

https://user-images.githubusercontent.com/7427545/196031804-d182757e-262a-4622-90fd-c2b6e8f5b4eb.mp4

**Comment by emilioicai** [comment]
@AlexanderEggers I tested windows support during the weekend and everything looked good on my end. on the other hand, I got stuck running `yarn fabric:android` as I kept getting build errors. Do you know if there's any kind of specific configuration I should have for it to run properly?

**Comment by matinzd** [comment]
Android is crashing for me:

```

FATAL EXCEPTION: main
Process: com.airbnb.android.lottie.rnexample, PID: 22894
java.lang.AssertionError: Method overloading is unsupported: com.facebook.react.uimanager.events.RCTModernEventEmitter#receiveEvent
	at com.facebook.react.bridge.JavaScriptModuleRegistry$JavaScriptModuleInvocationHandler.<init>(JavaScriptModuleRegistry.java:66)
	at com.facebook.react.bridge.JavaScriptModuleRegistry.getJavaScriptModule(JavaScriptModuleRegistry.java:41)
	at com.facebook.react.bridge.CatalystInstanceImpl.getJSModule(CatalystInstanceImpl.java:464)
	at com.facebook.react.bridge.ReactContext.getJSModule(ReactContext.java:160)
	at com.airbnb.android.react.lottie.LottieAnimationViewManager.sendOnAnimationFinishEvent(LottieAnimationViewManager.kt:56)
	at com.airbnb.android.react.lottie.LottieAnimationViewManager.access$sendOnAnimationFinishEvent(LottieAnimationViewManager.kt:30)
	at com.airbnb.android.react.lottie.LottieAnimationViewManager$createViewInstance$1.onAnimationEnd(LottieAnimationViewManager.kt:86)
	at android.animation.Animator$AnimatorListener.onAnimationEnd(Animator.java:552)
	at com.airbnb.lottie.utils.BaseLottieAnimator.notifyEnd(BaseLottieAnimator.java:74)
	at com.airbnb.lottie.utils.LottieValueAnimator.endAnimation(LottieValueAnimator.java:221)
	at com.airbnb.lottie.LottieDrawable.playAnimation(LottieDrawable.java:593)
	at com.airbnb.lottie.LottieDrawable.lambda$playAnimation$0$com-airbnb-lottie-LottieDrawable(LottieDrawable.java:578)
	at com.airbnb.lottie.LottieDrawable$$ExternalSyntheticLambda0.run(Unknown Source:2)
	at com.airbnb.lottie.LottieDrawable.setComposition(LottieDrawable.java:310)
	at com.airbnb.lottie.LottieAnimationView.setComposition(LottieAnimationView.java:550)
	at com.airbnb.lottie.LottieAnimationView$$ExternalSyntheticLambda0.onResult(Unknown Source:4)
	at com.airbnb.lottie.LottieTask.notifySuccessListeners(LottieTask.java:148)
	at com.airbnb.lottie.LottieTask.lambda$notifyListeners$0$com-airbnb-lottie-LottieTask(LottieTask.java:136)
	at com.airbnb.lottie.LottieTask$$ExternalSyntheticLambda0.run(Unknown Source:2)
	at android.os.Handler.handleCallback(Handler.java:790)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loop(Looper.java:164)
	at android.app.ActivityThread.main(ActivityThread.java:6494)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
```

**Comment by AlexanderEggers** [comment]
@matinzd Thanks for testing. I will try to have look in the next days. I think I have an idea how to fix that issue.

**Comment by AlexanderEggers** [comment]
@matinzd I did a bit more testing and I wasn't able to reproduce your issue. The only thing I found were duplicated .so files when building fabric-android. I resolved that as part of my most recent commits.

**Comment by mklb** [comment]
No idea if it's because of react-native 0.70.4 and new build architecture but my Lottie animations do not work on Android anymore after upgrading RN. They work on iOS though. 

Prop `autoPlay` does not work, calling something like `this.animationRef.current.play()` does not help as well.

lottie-react-native 5.1.4

**Comment by matinzd** [comment]
> No idea if it's because of react-native 0.70.4 and new build architecture but my Lottie animations do not work on Android anymore after upgrading RN. They work on iOS though.
> 
> Prop `autoPlay` does not work, calling something like `this.animationRef.current.play()` does not help as well.
> 
> lottie-react-native 5.1.4

Please open a separate issue for this.

**Comment by capezzbr** [comment]
Hey, this looks awesome, great job @AlexanderEggers  👏  How can we help move this PR forward? 🙏

**Comment by AlexanderEggers** [comment]
@alfonsocj The android integration still needs a bit of testing to verify that this implementation is actually stable. The iOS part is still missing and unfortunately something I haven't had the time to get started with.

**Comment by alfonsocj** [comment]
@AlexanderEggers thanks for the update! Are you planning on having a separate PR for iOS? I can try to tackle iOS based on the codegen work you've already done here. I have some experience adding fabric support for iOS as I did that for [react-native-datetimepicker](https://github.com/react-native-datetimepicker/datetimepicker/commit/8df590b51e80dfa5cd81bb5a4281de2f6ce139e5)

**Comment by BloodyMonkey** [comment]
Is there a temporary workaround ?
Our project is broken after upgrading RN

**Comment by alfonsocj** [comment]
@BloodyMonkey, the new architecture is optional. If you are blocked by this you can disable the new architecture (which is disabled by default).

**Comment by AlexanderEggers** [comment]
@alfonsocj I think we should get this PR merged first before we start any real iOS work - just to avoid any larger merge conflicts. I will probably have some time in around 2 weeks to get started or help on the iOS work. If you keen to get started with that already in a differeny PR, that would be great 🙂

**Review by matinzd — APPROVED** [review]
I was on vacation for a week but I tested the latest changes with both Android and iOS. It seems that the issue with Android is fixed and we can merge this to avoid any further conflict.

/cc @emilioicai

**Comment by norbusonam** [comment]
When will this be released?

**Comment by alfonsocj** [comment]
If there are no other features or fixes in master, I think we should wait until the iOS support for the new architecture gets merged. @emilioicai what do you think?

**Comment by emilioicai** [comment]
agreed, this kinda blocks fixes for other issues but I think we better wait for that merge. That gives us more time to test this one

**Comment by matinzd** [comment]
I am going to test this tonight. @emilioicai

**Comment by norbusonam** [comment]
do you have an idea of how long it will be until a next release? @alfonsocj @emilioicai

**Comment by alfonsocj** [comment]
@norbusonam I think we are waiting on fixing these issues before releasing it:
- https://github.com/lottie-react-native/lottie-react-native/issues/963
- https://github.com/lottie-react-native/lottie-react-native/issues/961

I don't have a lot of time right now to look at those. I believe @matinzd will be looking at those when he has some time.

If you have time to solve those, contributions are welcome!

**Comment by richierich1508** [comment]
Do you have an idea how long we need to wait for the new architecture (fabric) support in ios? @alfonsocj @emilioicai

**Comment by matinzd** [comment]
Hi @richierich1508,
As you may or may not know the community is trying to move to new architecture and it will take time for all libraries out there to adopt the new architecture. Most developers are also testing and benchmarking new architecture and I don't think many apps out there do have a new architecture turned on. I would suggest you turn it off for now and start using it in old architecture and when we were ready for release, you can start using it which we do not recommend it for the production track.

**Pull request merged**
